xarray/plot/plot.py

```diff
@@ -26,8 +30,132 @@
     get_axis,
     import_matplotlib_pyplot,
     label_from_attrs,
+    legend_elements,
 )
 
+# copied from seaborn
+_MARKERSIZE_RANGE = np.array([18.0, 72.0])
+
+
+def _infer_meta_data(darray, x, z, hue, hue_style, size):
+    def _determine_array(darray, name, array_style):
+        """Find and determine what type of array it is."""
+        array = darray[name]
+        array_is_numeric = _is_numeric(array.values)
+
+        if array_style is None:
+            array_style = "continuous" if array_is_numeric else "discrete"
+        elif array_style not in ["discrete", "continuous"]:
+            raise ValueError(
+                f"The style '{array_style}' is not valid, "
+                "valid options are None, 'discrete' or 'continuous'."
+            )
+
+        array_label = label_from_attrs(array)
+
+        return array, array_style, array_label
+
+    # Add nice looking labels:
+    out = dict(ylabel=label_from_attrs(darray))
+    out.update(
+        {
+            k: label_from_attrs(darray[v]) if v in darray.coords else None
+            for k, v in [("xlabel", x), ("zlabel", z)]
+        }
+    )
+
+    # Add styles and labels for the dataarrays:
+    for type_, a, style in [("hue", hue, hue_style), ("size", size, None)]:
+        tp, stl, lbl = f"{type_}", f"{type_}_style", f"{type_}_label"
+        if a:
+            out[tp], out[stl], out[lbl] = _determine_array(darray, a, style)
+        else:
+            out[tp], out[stl], out[lbl] = None, None, None
+
+    return out
+
+
+# copied from seaborn
+def _parse_size(data, norm, width):
```

Comment:
Is this copied from `dataset_plot.py`? Just checking... we could move it to `plot/utils.py` when you refactor the dataset scatter plot.

Reply:
It was indeed copied, but it's been modded a little I think. I mostly kept it here for debugging purposes.